core/deployment/src/main/java/io/quarkus/runner/RuntimeClassLoader.java

```diff
@@ -207,11 +206,15 @@ public InputStream getResourceAsStream(String nm) {
         Path classLoc = getClassInApplicationClassPaths(name);
 
         if (classLoc != null) {
-            CompletableFuture<Class<?>> res = new CompletableFuture<>();
-            Future<Class<?>> loadingClass = loadingClasses.putIfAbsent(name, res);
+            LoadingClass res = new LoadingClass(new CompletableFuture<>(), Thread.currentThread());
+            LoadingClass loadingClass = loadingClasses.putIfAbsent(name, res);
             if (loadingClass != null) {
+                if (loadingClass.initiator == Thread.currentThread()) {
+                    throw new LinkageError(
+                            "Load caused recursion in RuntimeClassLoader, this is a Quarkus bug loading class: " + name);
+                }
                 try {
-                    return loadingClass.get();
+                    return loadingClass.value.get();
```

Comment:
Maybe it would be nice to use the `get` method that accepts a timeout here to definitely avoid hanging?